### PR TITLE
Add configuration option to disable the bot

### PR DESCRIPTION
**Pull request by nicoroeser**
When the option is enabled, the scheduled tasks will do nothing.

This option can be useful, for example, if you like to disable resource
consumption by the crawler on your development systems, and like to do
this by means of a configuration setting instead of by disabling the
scheduled tasks.

### CODE REVIEW

**Comment by mattporritt** [comment]
Hi @nicoroeser,
This seems like an anti-feature to me.  I'm not sure I can see the advantage here over just disabling the related scheduled tasks.  I'll get a second opinion from the team on this.

@brendanheywood can you please offer your thoughts here.

Cheers,
Matt P

**Comment by brendanheywood** [comment]
Yeah I don't see a ton of value in having two settings that do the same thing. In other plugins we might add a disable flag but this is because it can be triggered by things other than cron like hooks. But that rule doesn't really apply here.

That said I can see some value in having the existing setting mirrored here as an admin convenience. I've wanted that for other plugins as well, but I'm still on the fence whether this is a good pattern to start. ie it simply gets and sets the same internal cron disable flag that you would see in the task detail edit form. If you really wanted it then I'd suggest you go down that path. Even better and simpler instead you could just add a convenience link the html at the top of this page which deep links into the existing core task editing page:

/admin/tool/task/scheduledtasks.php?action=edit&task=tool_crawler%5Ctask%5Ccrawl_task

**Comment by abias** [comment]
Hi @mattporritt and @brendanheywood ,

as I was the one who asked @nicoroeser to implement this feature, I would like to add some background information.

First, I agree that it does not make sense at first sight to have a kill switch for a plugin when there's already the possibility to disable the scheduled task. But the main difference is that you can set this kill switch in a Moodle instance's config.php, but you can't disable a scheduled task in config.php.

We have several testing and development instances which are mirrored (code, DB and Moodledata) from our production system. The only difference between the testing and development instances compared to the production system is that there are different settings in config.php regarding filesystem paths, URLs and database credentials and, most importantly, we have disabled or crippled all features which talk to backend / external systems or which would consume too many resources on testing instances. Disabling or crippling these features in config.php is important because a) config.php persists between resetting these systems and b) settings in config.php can't be overridden in the Moodle web interface by mistake or by a nosy admin.

Having said that, tool_crawler, is such a tool which just would waste resources on testing instances and thus we want to disable it hardly in config.php.

Currently, there isn't a possibility to disable a scheduled task in config.php.
I remember a discussion in Moodle Dev Chat about this request, but there even isn't a tracker issue about this.

As another approach, we have discussed disabling scheduled tasks with local_datacleaner in https://github.com/catalyst/moodle-local_datacleaner/issues/62, but this also got stuck and also would not _force_ disable the scheduled task.

Summing up, for our usage scenario, we would really appreciate if you could merge this kill switch feature at least as an intermedia solution.

As a side comment, having a kill switch could also improve admin usability. While a plugin's settings page is really obvious and on the spot, the plugin's scheduled task is rather hidden and probably unknown by not-so-experienced admins.

Cheers,
Alex

**Comment by mattporritt** [comment]
HI @abias ,
Thanks for the extra information, I'll discuss this more with the team internally and let you know.
Cheers,
Matt P

**Comment by brendanheywood** [comment]
@abias I completely agree with the business use case. Before you posted this last comment I was going to suggest this one below as I remember talking about it an assumed we'd actually done it:

https://github.com/catalyst/moodle-local_datacleaner/issues/62

Actually I think this IS implemented, at least partially (disables but doesn't force):

https://github.com/catalyst/moodle-local_datacleaner/tree/master/cleaner/scheduled_tasks

I assume you will also want to disable many other cron tasks not just this one, a generic solution either in the data cleaner or in config.php overrides is the better solution. There was not a tracker for the latter so I've made one:

https://tracker.moodle.org/browse/MDL-65843

We have more control over the datacleaner so I'd test that and get it over the line first if it isn't complete

**Comment by mattporritt** [comment]
Hi @brendanheywood and @abias ,
I've updated this branch and travis should now also pass.

Can you please let me know what is the next step on this request?

Cheers,
Matt P

**Comment by brendanheywood** [comment]
I'm not really opposed to it, I think the other solutions are better - but they are complimentary so I'll merge this in
